src/Opaleye/TH.hs

```diff
@@ -0,0 +1,652 @@
+        updateExp <- [e|liftIO $ Prelude.head <$> runUpdateReturning conn $(return $ VarE $ mkName $ makeTablename t) (\_ -> toWrite $ constant model) (\r -> ($(return $ VarE $ mkName pField) r) .== (constant $  $(return $ VarE $ mkName pField) model)) id |]
+        deleteExp <- [e|liftIO $ runDelete conn $(return $ VarE $ mkName $ makeTablename t) (\r -> ($(return $ VarE $ mkName pField) r) .== (constant $  $(return $ VarE $ mkName pField) model)) |]
+        let pat = [VarP $ mkName "conn", VarP $ mkName "model"]
+        let insertFunc = FunD (mkName "insertModel") [Clause pat (NormalB insertExp) convertToPgWrite]
+        let updateFunc = FunD (mkName "updateModel") [Clause pat (NormalB updateExp) convertToPgWrite]
+        let deleteFunc = FunD (mkName "deleteModel") [Clause pat (NormalB deleteExp) []]
+        return [InstanceD Nothing [] (AppT (ConT $ mkName "DbModel") (ConT $ mkName r)) [insertFunc, updateFunc, deleteFunc]]
+      where
+        makeConversionFunction :: Q [Dec]
+        makeConversionFunction = do
+          let
+            pgReadType = ConT $ (mkName $ makePGReadTypeName r)
+            pgWriteType = ConT $ (mkName $ makePGWriteTypeName r)
+            conversionFunctionSig = SigD (mkName "toWrite") $ AppT (AppT ArrowT pgReadType) pgWriteType
+            conversionFunction = FunD (mkName "toWrite") [Clause [makePattern] (NormalB conExp) []]
+          return $ [conversionFunctionSig, conversionFunction]
+          where
+            conExp :: Exp
+            conExp = foldl AppE (ConE $ mkName $ r) $ getFieldExps
+            getFieldExps :: [Exp]
+            getFieldExps = zipWith getFieldExp fieldInfos [1..]
+              where
+              getFieldExp :: ColumnInfo -> Int -> Exp
+              getFieldExp ci ix = if (columnDefault ci) then (AppE (ConE $ mkName "Just") mkVarExp) else mkVarExp
+                where
+                  mkVarExp :: Exp
+                  mkVarExp = VarE $ mkName $ "a" ++ (show ix)
+            makePattern :: Pat
+            makePattern = ConP (mkName $ r) $ VarP <$> (mkName.(\x -> ('a':x)).show) <$>  take (Prelude.length fieldInfos) [1..]
+        getPrimaryKeyField :: String -> String -> (Maybe String)
+        getPrimaryKeyField tname modelName = case (filter (\ci -> columnPrimary ci)) fieldInfos of
+            [primaryField] -> Just $ makeFieldName modelName (columnName primaryField)
+            _ -> Nothing
+  
+makeTablename :: String -> String
+makeTablename t = t ++ "Table"
+
+makePGReadTypeName :: String -> String
+makePGReadTypeName tn = tn ++ "PGRead"
+
+makePGWriteTypeName :: String -> String
+makePGWriteTypeName tn = tn ++ "PGWrite"
+
+makeAdapterName :: String -> String
+makeAdapterName tn = 'p':tn
+
+makeOpaleyeModels :: (ConnectInfo, Options) -> Q [Dec]
+makeOpaleyeModels (ci, op) = makeOpaleyeModels' (ci, op, [])
+
+makeOpaleyeModels' :: Env -> Q [Dec]
+makeOpaleyeModels' env = fst <$> runStateT (do
```

Comment:
>Perhaps collect all the "operations" in a large ADT first, then filter out the duplicates,

I am not sure if that is possible in this case, because we are using the Control.Lens functions for generating the actual lenses. These functions are the ones that is creating the classes (along with the instances). So if you have 3 tables, A, B and C, and B and C have columns with same name, say "createdDate", you can collect them and filter out duplicates, but when you are calling the makeLenses function for A, B and C in sequence, you need to indicate to the call for B that the HasCreatedDate type class is not generated, so create the class and for the call for C, that "createdDate" has been already created, so skip the class generation.

Normally, this would be fine, since the makeLenses function automatically skips Classes for existing classes. But in our case, the lenses and classes for B has not been inserted into the scope, when we making the lenses for C, which requires us to keep track of the generated classes...

All of this means maintaining a state, by using SateT or by doing it manually, which I think will be a lot more clumsier and error prone than using StateT.